integration-tests/src/codegen/codegen_documentation.rs

```diff
@@ -0,0 +1,119 @@
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// subxt is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with subxt.  If not, see <http://www.gnu.org/licenses/>.
+
+use regex::Regex;
+use subxt_codegen::{
+    DerivesRegistry,
+    RuntimeGenerator,
+};
+
+fn metadata_docs() -> Vec<String> {
+    // Load the runtime metadata downloaded from a node via `test-runtime`.
+    let bytes = test_runtime::METADATA;
+    let meta: frame_metadata::RuntimeMetadataPrefixed =
+        codec::Decode::decode(&mut &*bytes).expect("Cannot decode scale metadata");
+    let metadata = match meta.1 {
+        frame_metadata::RuntimeMetadata::V14(v14) => v14,
+        _ => panic!("Unsupported metadata version {:?}", meta.1),
+    };
+
+    // Inspect the metadata types and collect the documentation.
+    let mut docs = Vec::new();
+    for ty in metadata.types.types() {
+        docs.extend_from_slice(ty.ty().docs());
+    }
+
+    for pallet in metadata.pallets {
+        if let Some(storage) = pallet.storage {
+            for entry in storage.entries {
+                docs.extend(entry.docs);
+            }
+        }
+        // Note: Calls, Events and Errors are deduced directly to
+        // PortableTypes which are handled above.
+        for constant in pallet.constants {
+            docs.extend(constant.docs);
+        }
+    }
+    // Note: Extrinsics do not have associated documentation, but is implied by
+    // associated Type.
+
+    docs
+}
+
+fn generate_runtime_interface() -> String {
+    // Load the runtime metadata downloaded from a node via `test-runtime`.
+    let bytes = test_runtime::METADATA;
+    let metadata: frame_metadata::RuntimeMetadataPrefixed =
+        codec::Decode::decode(&mut &*bytes).expect("Cannot decode scale metadata");
+
+    // Generate a runtime interface form the provided metadata.
+    let generator = RuntimeGenerator::new(metadata);
+    let item_mod = syn::parse_quote!(
+        pub mod api {}
+    );
+    let derives = DerivesRegistry::default();
+    generator.generate_runtime(item_mod, derives).to_string()
+}
+
+fn interface_docs() -> Vec<String> {
+    // Generate the runtime interface from the node's metadata.
+    // Note: the API is generated on a single line.
+    let runtime_api = generate_runtime_interface();
+
+    // Documentation lines have the following format:
+    //    # [ doc = "Upward message is invalid XCM."]
+    // Given the API is generated on a single line, the regex matching
+    // must be lazy hence the `?` in the matched group `(.*?)`.
```

Comment:
I guess if any doc string contains `"]` we'd hit an issue here?

perhaps it could also be:

```
let re = Regex::new(r#"\# \[doc = "[^\n]*"\]"#).unwrap();
```

To just explicitly forbid actual newlines from appearing and force the match to be on one line, for a bit of added reliability?

Reply:
Had a try printing all the documentation pulled from the generated file 
Seems to be also handling the cases for extra `]`:

```
			The [event](https://docs.substrate.io/v3/runtime/events-and-errors) emitted
			by this pallet.
# <weight>
- `O(1)`.
```

Although not entirely accurate: https://regex101.com/r/TtWQqO/1 . Seems to handle multiple `]` on the same documentation line.
The `^\n` didn't seem to match tho 🤔